lib/Doctrine/DBAL/Driver/SQLite3/SQLite3Connection.php

```diff
@@ -0,0 +1,216 @@
+ */
+
+namespace Doctrine\DBAL\Driver\SQLite3;
+
+use Doctrine\DBAL\Driver\Connection;
+use Doctrine\DBAL\Driver\ServerInfoAwareConnection;
+
+/**
+ * SQLite3 implementation of the Connection interface.
+ *
+ * @since 2.6
+ * @author Ben Morel <benjamin.morel@gmail.com>
+ * @author Bill Schaller <bill@zeroedin.com>
+ */
+class SQLite3Connection extends SQLite3Abstract implements Connection, ServerInfoAwareConnection
+{
+    /**
+     * @param string      $filename
+     * @param int|null    $flags
+     * @param string|null $encryption_key
+     *
+     * @throws SQLite3Exception
+     */
+    public function __construct($filename, $flags = null, $encryption_key = null)
+    {
+        try {
+            if ($flags === null) {
+                $this->sqlite3 = new \SQLite3($filename);
+            } elseif ($encryption_key === null) {
+                $this->sqlite3 = new \SQLite3($filename, $flags);
+            } else {
+                $this->sqlite3 = new \SQLite3($filename, $flags, $encryption_key);
+            }
+
+            $this->sqlite3->enableExceptions(true);
+        } catch (\Exception $e) {
+            throw $this->wrapDriverException($e);
+        }
+    }
+
+    /**
+     * Returns the underlying SQLite3 object.
+     *
+     * @return \SQLite3
+     */
+    public function getSQLite3()
+    {
+        return $this->sqlite3;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function prepare($prepareString)
+    {
+        try {
+            $statement = @ $this->sqlite3->prepare($prepareString);
+        } catch (\Exception $e) {
+            throw $this->wrapDriverException($e);
+        }
+
+        $this->throwExceptionOnError();
+
+        return new SQLite3Statement($this->sqlite3, $statement);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function query()
+    {
+        $query = func_get_arg(0);
+
+        $statement = $this->prepare($query);
+        $statement->execute();
+
+        return $statement;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function quote($input, $type = \PDO::PARAM_STR)
+    {
+        return "'" . \SQLite3::escapeString($input) . "'";
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function exec($statement)
+    {
+        try {
+            $this->sqlite3->exec($statement);
+        } catch (\Exception $e) {
+            throw $this->wrapDriverException($e);
+        }
+
+        $this->throwExceptionOnError();
+
+        return $this->sqlite3->changes();
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function lastInsertId($name = null)
+    {
+        try {
+            return $this->sqlite3->lastInsertRowID();
+        } catch (\Exception $e) {
+            throw $this->wrapDriverException($e);
+        }
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function beginTransaction()
+    {
+        try {
+            $this->sqlite3->exec('BEGIN');
+        } catch (\Exception $e) {
```

Comment:
That's a great point, @deeky666 - the drivers are extremely inconsistent. Even DBAL\Connection has void return types for commit, rollback, and beginTransaction, where the contract has return true on success, false on error.

The question is: what is the better behavior? Return true/false on error or raise an exception? I'd argue that raising exceptions is the expected behavior, especially within the ORM. If you look at usages of beginTransaction, commit, and rollback, the ORM doesn't use the return values on those methods at all. It simply expects an exception to be thrown if there's a problem.

Using true/false return values is pretty fragile IMO. It relies on the value being returned throughout the call chain, and it also forces consumers to either create code paths to marshal error information via special getErrorInfo methods throughout all consumers, or via direct access to the underlying statement (bad).

I'd say throwing an exception is much cleaner, because even if the exception is wrapped multiple times, if they're properly handled, the $previous exception will chain back to the originally thrown one. It also allows us to abstract over the oddities of driver return values. Some functions may return true or void on error where others return false. Using exceptions throughout the DBAL driver stack lets us reason a bit more clearly about what is actually happening.